detect broken attachments URLs in posts content

kariae opened the pull request:
This PR adds `get_broken_attachment_urls_from_posts` method to the `Attachments` logic class, it returns broken URLs indexed by post IDs. 

The function takes into consideration the usage of [s3_uploads](https://github.com/humanmade/S3-Uploads) plugin which switches the image URL on the fly when we're rendering the post frontend (with `the_content` filter).

This can be used in something like this:

```php
use \NewspackCustomContentMigrator\MigrationLogic\Attachments as AttachmentsLogic;

// ...

$result = $this->attachments_logic->get_broken_attachments_urls_from_posts_content();
// $result = $this->attachments_logic->get_broken_attachments_urls_from_posts_content( [ 443372, 440745 ] );
foreach($result as $post_id => $broken_urls) {
	$post = get_post($post_id);
	$updated_post_content = $post->post_content;

	foreach($result as $post_id => $broken_urls) {
		foreach($broken_urls as $broken_url) {
			$fixed_url = $this->get_fixed_url($broken_url);
			if($fixed_url !== $broken_url) {
				$updated_post_content = str_replace($broken_url, $fixed_url, $updated_post_content);
			}
		}	
	}	

	if($updated_post_content !== $post->post_content) {
		// update post
	}
}
```

iuravic reviewed (changes requested):
@kariae, this is amazing! ✋ It does just what we said it should do.

A question -- could we add two very minor things, which would probably make this fully reusable:

1. when we fetch the [img srcs](https://github.com/Automattic/newspack-custom-content-migrator/pull/187/files#diff-4e9c34a4a7c25e3e3e72e9f64ea3278ec7ce12b10f1c01a3e6f1f6ce086e491fR110), we should probably check if the hostname is local (e.g. if an image is hard-coded and displayed from www.google.com, we'll want to skip those)

2. could we add an argument/flag which would control if this command should use S3 hostnames, or local ones? That could control if the [src here](https://github.com/Automattic/newspack-custom-content-migrator/pull/187/files#diff-4e9c34a4a7c25e3e3e72e9f64ea3278ec7ce12b10f1c01a3e6f1f6ce086e491fR131) check imgs from S3 or local, and we could use it to check non-S3 sites too (and add this to our new PrelaunhSiteQAMigrator)

kariae commented on src/MigrationLogic/Attachments.php:
```diff
@@ -80,4 +80,82 @@ public function import_external_file( $path, $title = null, $caption = null, $de
+						$month        = prev( $exploded_url );
+						$year         = prev( $exploded_url );
+						$s3_url       = 'https://' . $bucket . ".s3.amazonaws.com/wp-content/uploads/$year/$month/$filename";
+
```
@eddiesshop when we do upload a media on WP it ends on the /uploads/year/month folder, is there a case where this can be changed? If yes, we can add more checks.

In general, I didn't put many checks in here, because this is the second check, and if doesn't work it means that probably the media URL format is not the default one, and maybe we'll need to override the whole thing.

eddiesshop replied:
Totally true! I was just thinking about files that are uploaded to the root uploads folder. But I guess those would be out of scope?

kariae commented:
@iuravic I implemented the requested changes and added a command to switch local URLs to S3 URLs we used it to clean the Berkeleyside broken URLs. If you or @eddiesshop can do a final review before merging this, please.
